### PR TITLE
L2 refactor

### PR DESCRIPTION
**Pull request by kmokstad**
Includes an option (currently hardcoded) to enforce the end-point values in the L2-projection.
Not very well tested - only in 1D, so this is WIP.

### CODE REVIEW

**Comment by akva2** [comment]
missing: ASMu2Dmxrecovery.C.

worth considering: ASMLRSpline.C, edgeL2projection (and the equivalent faceL2projection in https://github.com/OPM/IFEM/pull/158) and the related duplication of code in ASMxyD implementations of updateDirichlet.

**Comment by kmokstad** [comment]
I know, did not consider the mixed patches yet, until we know that this is actually how we will do it.
ASMbase::globalL2projection cannot be used for mixed, due to the different arrangement of the field variables.

The suggested refactor of the LR-L2-projection is also worth considering, but that is for projecting explicit functions for non-homogeneous boundary conditions, not the secondary solution as here.

**Comment by akva2** [comment]
Yes. It could possibly be unified through fieldfunc, though there need to be substantial code simplicity to be had to go that way as there would be some overhead.

**Comment by kmokstad** [comment]
jenkins build this with downstreams please

**Comment by kmokstad** [comment]
jenkins build this with downstreams please

File: src/ASM/LR/ASMu2Dmx.C [hunk 1]
```diff
@@ -745,20 +745,19 @@ void ASMu2Dmx::generateThreadGroups (const Integrand& integrand, bool silence,
...
   // TODO: Support for div-compatible
   int p1 = 0;
-  for (size_t i = 1; i <= m_basis.size(); ++i) {
-    if (this->getBasis(i)->order(0) > p1)
-      p1 = this->getBasis(i)->order(0), threadBasis = i;
+  for (size_t i = 1; i <= m_basis.size(); ++i)
+    if (this->getBasis(i)->order(0) > p1) {
+      threadBasis = this->getBasis(i);
+      p1 = threadBasis->order(0);
   }
```
Comment (akva2) [inline src/ASM/LR/ASMu2Dmx.C]: wrong indent

**Comment by kmokstad** [comment]
Only the last commit here is not merged yet. That one is put on hold as not sure yet if this is needed.

**Comment by akva2** [comment]
it could be used to simplify constrainEdge/constrainFace and clarify the role of ASMbase::updateDirichlet (only handles actual nodal constraints).

**Comment by akva2** [comment]
could you rebase please?

**Comment by kmokstad** [comment]
Most of the refactoring mentioned above was already done long time ago (unifying the globalL2projection methods into a ASMbase method, etc.). Then only thing remaining here is the option to enforce the end/corner point values. Therefore rewriting this branch with that only on top + some small doxy fixes + a potiential bugfix in L2FuncIntegrand::evaluate()  (see second commit).

**Comment by kmokstad** [comment]
jenkins build this with downstreams please

**Pull request merged**